Doing kNN by Scratch got me nowhere, so let's try it with sklearn instead!

In [2]:
import pandas as pd
names = ['Index', 'Rank','Song','Artist','Year','Lyrics','ids','Energy','Liveness','Tempo','Speechiness','Acousticness','Instrumentalness','Time Signature','Danceability','Key','Duration (ms)','Loudness','Valence','Quartile']
df = pd.read_csv('data/FINAL_cleaned_output_quartile.csv', header=None, names=names)
df.head()

Index     Rank          Song                            Artist  Year  \
0  Index  �_�Rank          Song                            Artist  Year   
1      0        1       breathe                        faith hill  2000   
2      1        2        smooth      santana featuring rob thomas  2000   
3      2        3   maria maria  santana featuring the product gb  2000   
4      3        4  i wanna know                               joe  2000   

                                              Lyrics                     ids  \
0                                             Lyrics                     ids   
1   i can feel the magic floating in the air bein...  0KsB4TwgATg88aXCMBoO3Y   
2   man its a hot one like seven inches from the ...  31x4wWBghN7f5PiEZUmcAW   
3   ladies and gents turn up your sound systems t...  3XKIUb7HzIF1Vu9usunMzc   
4   yeah oh yeah alright oh oh ohits amazing how ...  5FVd6KXrgO9B3JPmC8OPst   

   Energy  Liveness    Tempo  Speechiness  Acousticness  Instrumentalness  \
0  energy  liveness    tempo  speechiness  acousticness  instrumentalness   
1   0.728      0.14  125.984       0.0408        0.0259           0.00025   
2   0.848     0.658  161.049       0.0538         0.214          1.84E-05   
3   0.601    0.0348   97.911        0.126        0.0406           0.00201   
4   0.532     0.217    85.03       0.0323         0.186          0.000263   

   Time Signature  Danceability  Key Duration (ms)  Loudness  Valence  \
0  time_signature  danceability  key   duration_ms  loudness  valence   
1               4         0.741   11        207630    -5.559    0.703   
2               4         0.471    6        170200    -5.355    0.332   
3               4         0.777    2        261973    -5.931     0.68   
4               4         0.548    5        272394    -7.596    0.405   

   Quartile  
0  quartile  
1         1  
2         1  
3         1  
4         1

In [3]:
import numpy as np
from sklearn.cross_validation import train_test_split
# [7,8,9,10,11,12,13,14,15,16,17,18]
# X = np.array(df.ix[1:, 7:18])
# Energy, Liveness, Tempo
# X = np.array(df.ix[1:, 7:9])
# Speechiness, Acousticness, Instrumentalness
# X = np.array(df.ix[1:, 10:12])
# Danceability and loudness
X = np.array(df.ix[1:, 7:18])
X = np.asfarray(X,float)
#count = 0
#for x in X:
#    count += 1
#    for y in x:
#    if np.isnan(x[0]):
#    print count
#     print x
#    print np.isfinite(x)
#X = np.asfarray(X,float)
y = np.array(df.ix[1:, 19])
y = np.asfarray(y,float)
#print y
#print np.isfinite(y)
#for Y in y:
#    count += 1
#    if np.isnan(Y):
#        print count
#        print Y
#X = np.delete(X, (627), axis=0)
#X = np.delete(X, (361), axis=0)
#X = np.delete(X, (374), axis=0)
#X = np.delete(X, (409), axis=0)
#X = np.delete(X, (587), axis=0)
#X = np.delete(X, (706), axis=0)
#y = np.delete(y, (627), axis=0)
#y = np.delete(y, (361), axis=0)
#y = np.delete(y, (374), axis=0)
#y = np.delete(y, (409), axis=0)
#y = np.delete(y, (587), axis=0)
#y = np.delete(y, (706), axis=0)
print X.shape
print y.shape
#print np.isnan(X).any()
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.33,random_state=42)
#X_train = np.delete(X_train, (705), axis=0)
#y_train = np.delete(y_train, (705), axis=0)
#print np.isnan(X_train).any()
#for xt in X_train:
#    count += 1
#    if(np.isnan(xt).any()):
#        print count
#        print xt
#print np.isnan(y_train).any()

(1274, 11)
(1274,)


/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:10: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  # Remove the CWD from sys.path while we load stuff.


In [4]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train, y_train)
pred = knn.predict(X_test)
#print pred
print accuracy_score(y_test, pred) * 100

25.653206650831358


## The Current Analysis - Predicting Quartile

### kNN Accuracy Scores for Given k Values

#### Using ALL Audio Features
* 1 -> 22.8029 %
* 2 -> 23.9905 %
* 3 -> 23.5154 %
* 4 -> 23.7530 %
* 5 -> 23.2779 %
* 6 -> 25.4157 %
* 7 -> 22.5653 %
* 8 -> 25.4157 %
* 9 -> 25.1781 %
* 10 -> 25.6532 %

So the accuracy is better. Not great, but better. HOWEVER, we can also check to see how often the classifier classified within 1 of the actual Quartile of the song in question. 

In [9]:
#print len(pred)
#print len(y_test)
#print y_test.shape
#print y_test
differences = list()
for q in range(len(y_test)):
    differences.append(y_test[q]-pred[q])
acc0 = 0
for w in differences:
    if w == 0:
        acc0 += 1
print "Completely Accurate",(acc0/float(len(pred)))*100, "%"
acc1 = 0
for w in differences:
    if w == -1:
        acc1 += 1
print "One Above",(acc1/float(len(pred)))*100, "%"
acc2 = 0
for w in differences:
    if w == 1:
        acc2 += 1
print "One Below",(acc2/float(len(pred)))*100, "%"
print "Sum of the Three", ((acc0+acc1+acc2)/float(len(pred)))*100, "%"

Completely Accurate 25.6532066508 %
One Above 16.1520190024 %
One Below 18.7648456057 %
Sum of the Three 60.5700712589 %


### kNN Accuracy within +/- One Quartile
* 1 -> 60.5701 %
* 2 -> 61.9952 %
* 3 -> 60.8076 %
* 4 -> 60.095 %
* 5 -> 61.0451 %
* 6 -> 61.2827 %
* 7 -> 58.1948 %
* 8 -> 59.6200 %
* 9 -> 61.5202 %
* 10 -> 60.5701 %

Wow. That was... much better! We can predict within +/- 1 quartile at least 60% of the time, which means that generally speaking while we may not always be able to predict exactly what rank a song would be, we can more often predict its quartile and we can more often than not predict within 1 quartile of its actual value.

In [6]:
quart1count = 0
diffquart1 = list()
quart2count = 0
diffquart2 = list()
quart3count = 0
diffquart3 = list()
quart4count = 0
diffquart4 = list()
for q in range(len(y_test)):
    if y_test[q] == 1:
        quart1count += 1
        diffquart1.append(y_test[q]-pred[q])
    if y_test[q] == 2:
        quart2count += 1
        diffquart2.append(y_test[q]-pred[q])
    if y_test[q] == 3:
        quart3count += 1
        diffquart3.append(y_test[q]-pred[q])
    if y_test[q] == 4:
        quart4count += 1
        diffquart4.append(y_test[q]-pred[q])
accq1= 0
for h in diffquart1:
    if h == 0:
        accq1 += 1
print "Accurately Predicted as Quartile 1",(accq1/float(len(diffquart1)))*100, "%"
accq2= 0
for h in diffquart2:
    if h == 0:
        accq2 += 1
print "Accurately Predicted as Quartile 2",(accq2/float(len(diffquart2)))*100, "%"
accq3= 0
for h in diffquart3:
    if h == 0:
        accq3 += 1
print "Accurately Predicted as Quartile 3",(accq3/float(len(diffquart3)))*100, "%"
accq4= 0
for h in diffquart4:
    if h == 0:
        accq4 += 1
print "Accurately Predicted as Quartile 4",(accq4/float(len(diffquart4)))*100, "%"

Accurately Predicted as Quartile 1 38.4615384615 %
Accurately Predicted as Quartile 2 24.5098039216 %
Accurately Predicted as Quartile 3 16.9811320755 %
Accurately Predicted as Quartile 4 22.9357798165 %


So it seems that certain quartiles are predicted correctly more often than others. Songs in Quartile 1 are predicted correctly more often than songs in any other quartile, and songs in Quartile 3 are predicted correctly the least often.

* Q1: 38.4615%
* Q2: 24.5098%
* Q3: 16.9811%
* Q4: 22.9358%

In [8]:
accq1a= 0
for h in diffquart1:
    if h == -1:
        accq1a += 1
print "Predicted one Above Quartile 1",(accq1a/float(len(diffquart1)))*100, "%"
print "Dead on Or One Above", ((accq1a+accq1)/float(len(diffquart1)))*100, "%"

Predicted one Above Quartile 1 24.0384615385 %
Dead on Or One Above 62.5 %


## The Previous Analysis - Predicting Rank

### kNN Accuracy Scores for Given k Values - Cleaned Data

#### Using ALL Audio Features
* 1 -> 0.4751 %
* 2 -> 1.6627 %
* 3 -> 1.9002 %
* 4 -> 2.1378 %
* 5 -> 1.1876 %
* 6 -> 0.7126 %
* 7 -> 0.7126 %
* 8 -> 1.1876 %
* 9 -> 1.1876 %
* 10 -> 0.9501 %

#### Using Energy, Liveness, and Tempo
* 1 -> 0.2375 %
* 2 -> 1.6627 %
* 3 -> 1.6627 %
* 4 -> 2.1378 %
* 5 -> 1.6627 %
* 6 -> 1.6627 %
* 7 -> 1.4252 %
* 8 -> 1.1876 %
* 9 -> 1.1876 %
* 10 -> 1.4252 %

#### Using Speechiness, Acousticness, and Instrumentalness
* 1 -> 0.7126 %
* 2 -> 1.4252 %
* 3 -> 1.4252 %
* 4 -> 1.1876 %
* 5 -> 1.1876 %
* 6 -> 1.1876 %
* 7 -> 1.1876 %
* 8 -> 1.1876 %
* 9 -> 0.9501 %
* 10 -> 1.1876 %

#### Using Danceability and Loudness
* 1 -> 0.7126 %
* 2 -> 0.9501 %
* 3 -> 0.7126 %
* 4 -> 0.7126 %
* 5 -> 0.7126 %
* 6 -> 0.7126 %
* 7 -> 0.7126 %
* 8 -> 0.7126 %
* 9 -> 0.7126 %
* 10 -> 0.7126 %

Yeah. This went absolutely nowhere. I can't draw any meaningful conclusion from the above other than that kNN isn't useful in predicting rank based on audio features.

Ok, well, unless I can figure out what went horribly wrong with kNN, we'll try out Naive Bayes instead (ty Susie)

In [2]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
from sklearn.naive_bayes import BernoulliNB

In [8]:
print df.describe()

        Rank                        Song   Artist  Year  \
count   1276                        1276     1276  1276   
unique   101                        1276      822    17   
top       11  i write sins not tragedies  rihanna  2000   
freq      16                           1       16    95   

                                                   Lyrics  \
count                                                1251   
unique                                               1250   
top      yeah cmon i need a i want ai need a girl to r...   
freq                                                    2   

                           ids Energy Liveness    Tempo Speechiness  \
count                     1275   1275     1275     1275        1275   
unique                    1275    548      633     1236         673   
top     7Ie9W94M7OjPoZVV216Xus  0.726    0.109  125.014      0.0311   
freq                         1      9       15        3           6   

       Acousticness Instrumentalness Time 